src/when-contexts.ts

```diff
@@ -1,56 +1,56 @@
-import { TokenCursor } from './cursor-doc/token-cursor';
+import { Token } from './cursor-doc/lexer';
+import { equal as deep_equal } from './cursor-doc/model'
+import { TokenCursor, LispTokenCursor } from './cursor-doc/token-cursor';
 import * as docMirror from './doc-mirror'
 
-type CursorContext = 'calva-standard' | AltCursorContext;
-type AltCursorContext = 'string' | 'comment';
-const STRING_CONTEXT = 'calva:cursorInString';
-const COMMENT_CONTEXT = 'calva:cursorInComment';
-let lastContext: CursorContext = null;
+type CursorContext = LexerCursorContext | PreLexerCursorContext;
+type LexerCursorContext = 'calva:cursorInString' | 'calva:cursorInComment';
+type PreLexerCursorContext = 'calva:cursorAtStartOfLine' | 'calva:cursorAtEndOfLine'
+const allCursorContexts: CursorContext[] = ['calva:cursorInString', 'calva:cursorInComment', 'calva:cursorAtStartOfLine', 'calva:cursorAtEndOfLine']
+
+let lastContexts: CursorContext[] = [];
 
 export default function setCursorContextIfChanged(editor: vscode.TextEditor) {
     if (!editor || !editor.document || editor.document.languageId !== 'clojure') return;
 
-    const currentContext = determineCursorContext(editor.document, editor.selection.active);
-    setCursorContext(currentContext);
+    const currentContexts = determineCursorContexts(editor.document, editor.selection.active);
+    setCursorContexts(currentContexts);
 }
 
-function setCursorContext(currentContext: CursorContext) {
-    if (lastContext === currentContext) {
+function setCursorContexts(currentContexts: CursorContext[]) {
+    if (deep_equal(lastContexts, currentContexts)) {
         return;
     }
-    lastContext = currentContext;
-   
-    vscode.commands.executeCommand('setContext', COMMENT_CONTEXT, false);
-    vscode.commands.executeCommand('setContext', STRING_CONTEXT, false);
-
-    switch (currentContext) {
-        case 'calva-standard':
-            break;
-        case 'comment':
-            vscode.commands.executeCommand('setContext', COMMENT_CONTEXT, true);
-            break;
-        case 'string':
-            vscode.commands.executeCommand('setContext', STRING_CONTEXT, true);
-            break;
-        default:
-            const checkExhaustive: never = currentContext;
-    }
+    lastContexts = currentContexts;
+
+    allCursorContexts.forEach(context => {
+        vscode.commands.executeCommand('setContext', context, currentContexts.indexOf(context) > -1)
+    })
 }
 
-function determineCursorContext(document: vscode.TextDocument, position: vscode.Position): CursorContext {
+function determineCursorContexts(document: vscode.TextDocument, position: vscode.Position): CursorContext[] {
+    let contexts: CursorContext[] = [];
     const idx = document.offsetAt(position);
     const mirrorDoc = docMirror.getDocument(document);
     const tokenCursor = mirrorDoc.getTokenCursor(idx);
+    const line = document.lineAt(position);
+
+    if (position.character <= line.firstNonWhitespaceCharacterIndex) {
```

Comment:
I think vscode and Clojure might have different definitions of whitespace, (the comma.) You can use the token cursor instead here. So if the previous token is `ws` then `backwardWhiteSpace()` and then if the new position is on the previous row or the column is 0. Or something along those lines.

(I saw you said you're not done with this change yet, just throwing this in here since it is easy to forget about that comma.)

Reply:
I filed a feature request on vscode now. It really would be nice to solve this like you did there.

https://github.com/microsoft/vscode/issues/117174

Reply:
You're right  - hadn't thought of commas. Still something of a clojure rube really